pinot/net.py

```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
     learning, puts into a predicted distribution and calculates the
     corresponding divergence.
 
+
+    Attributes
+    ----------
+    representation: a `pinot.representation` module
+        the model that translates graphs to latent representations
+    output_regression: a `torch.nn.Moudle` or None,
+        if None, this will be set as a simple `Linear` layer that inputs
+        the latent dimension and output the number of parameters for
+        `self.distribution_class`
+    distribution_class: a `torch.distribution` object,
+        default=`torch.distributions.normal.Normal`
+        the class of distribution the model outputs
+    noise_model: either a string (
+        one of 
+            'homoschedastic',
+            'heteroschedastic',
+            'homoschedastic-fixed') 
+        or a function that transforms a set of parameters.
+
+
     """
 
     def __init__(
         self,
         representation,
-        parameterization,
+        output_regression=None,
         distribution_class=torch.distributions.normal.Normal,
-        param_transform=lambda x, y: (x, torch.exp(y) + 1e-5),
-        expectation_fn=lambda x, y: x,
+        noise_model='heteroschedastic',
     ):
+        
         super(Net, self).__init__()
         self.representation = representation
-        self.parametrization = parameterization
         self.distribution_class = distribution_class
-        self.param_transform = param_transform
-        self.expectation_fn = expectation_fn
+
+        # grab the last dimension of `representation`
+        regression_in_features = [
+                layer for layer in list(self.representation.modules())\
+                        if hasattr(layer, 'out_features')][-1].out_features
+
+        # TODO:
+        # make this less ugly
+        regression_out_features = len(distribution_class.arg_constraints)
+
+        if output_regression is None:
+            # make the output regression as simple as a linear one
+            # if nothing is specified
+            self._output_regression = torch.nn.ModuleList(
+                    [
+                        torch.nn.Linear(regression_in_features, 1)\
+                                for _ in range(regression_out_features)
+                    ])
+
+            def output_regression(theta):
```

Comment:
does it need that? imho if you build an nn.sequential with a modulelist object inside, i.e.

regressor_net = nn.Sequential(self._output_regression) 

you should be able to just call .forward() on that later and get back your list without needing a for loop through these modules.

Reply:
here we're using `torch.nn.ModuleList` and the `forward` function of it raises NotImplemented Error

Reply:
did you try first wrapping it into a sequential?

Reply:
i.e. ```regressor =  nn.Sequential(module_list_you_created)``` should work with .forward()

Reply:
it might need an identity layer pre-prended or so, let me know if it does not work out of the box.

Reply:
ok I think this is resolved, you indeed need that for loop